components/video-recorder/src/HeadlessVideoRecorder.tsx

```diff
@@ -0,0 +1,636 @@
+		},
+		[recording, handleDownload, showPreview],
+	);
+
+	const wait = (delayInMS: number) => {
+		return new Promise((resolve) => setTimeout(resolve, delayInMS));
+	};
+
+	const record = async (stream: MediaStream, lengthInMs: number) => {
+		const recorder = new MediaRecorder(stream, {
+			mimeType,
+			videoBitsPerSecond: 2500000,
+		});
+		mediaRecorderRef.current = recorder;
+		const data: Blob[] = [];
+
+		recorder.ondataavailable = (event) => {
+			data.push(event.data);
+		};
+
+		recorder.start();
+
+		timer.start(lengthInMs / 1000);
+		const recorded = await wait(lengthInMs);
+
+		if (recorder.state === "recording") {
+			recorder.stop();
+			timer.stop();
+		}
+		const stopped = new Promise((resolve, reject) => {
+			recorder.onstop = resolve;
+			recorder.onerror = (event) => {
+				timer.stop();
+				reject(event);
+			};
+		});
+
+		try {
+			await Promise.all([stopped, recorded]);
+			return data;
+		} catch (error) {
+			timer.stop();
+			console.log(error);
+		}
+	};
+
+	const stopPreviewStream = () => {
+		if (videoRef.current?.srcObject instanceof MediaStream) {
+			const stream = videoRef.current?.srcObject;
+			stream?.getTracks().forEach((track) => track.stop());
+		}
+	};
+
+	const onStopMediaStream = (recordedChunks: Blob[] | undefined) => {
+		const recorderBlob = new Blob(recordedChunks, { type: blobType });
+		const videoSrc = URL.createObjectURL(recorderBlob);
+
+		const extension = fileExtension;
+		const recordedFile = new File([recorderBlob], `video.${extension}`, {
+			type: extension,
+		});
+		setFile(recordedFile);
+		if (playbackRef.current) {
+			playbackRef.current.src = videoSrc;
+		}
+	};
+
+	const handleStartRecording = async () => {
+		try {
+			if (autoStop) {
+				setRecording("recording");
+				if (videoRef.current?.srcObject instanceof MediaStream) {
+					const recordedChunks: Blob[] | undefined = await record(
+						videoRef.current?.srcObject,
+						recordingTime,
+					);
+					onStopMediaStream(recordedChunks);
+					setRecording("recorded");
+					stopPreviewStream();
+				}
+			} else {
+				if (videoRef.current?.srcObject instanceof MediaStream) {
+					mediaRecorderRef.current = new MediaRecorder(
+						videoRef.current?.srcObject,
+						{
+							mimeType,
+							videoBitsPerSecond: 2500000,
+						},
+					);
+					const chunks: Blob[] = [];
+					mediaRecorderRef.current.ondataavailable = (e) => {
+						if (e.data.size > 0) {
+							chunks.push(e.data);
+						}
+					};
+					mediaRecorderRef.current.onstop = () => {
+						onStopMediaStream(chunks);
+					};
+					mediaRecorderRef.current.start();
+					timer.startFromZero();
+					setRecording("recording");
+				}
+			}
+		} catch (error) {
+			console.log(error);
+		}
+	};
+
+	const handleStopRecording = () => {
+		if (mediaRecorderRef.current && recording == "recording") {
+			mediaRecorderRef.current?.stop();
+			timer.stop();
+		}
+		stopPreviewStream();
+		setRecording("recorded");
+	};
+
+	const handleRetake = () => {
```

Comment:
Type of this function will be `()=> void`. This can be used when this type is required.